### PR TITLE
Claim Aware Ability

### PR DESCRIPTION
**Pull request by aliprogrammer69**
with this changes, module can grant claims

### CODE REVIEW

**Comment by max-lt** [comment]
Hi, thanks for the PR, I will review it as soon as I can

File: README.md [hunk 1]
```diff
@@ -38,6 +38,16 @@ server {
         auth_jwt on;
     }
 
+    location /secured-by-claim/ {
+        auth_jwt_claim claim_name claim_value;
+        auth_jwt on;
+    }
+
+    location /secured-by-claim-variable/ {
+        auth_jwt_claim claim_name $Variable;
```
Comment (max-lt) [inline README.md:47]: lowercase Variable -> variable

File: src/ngx_http_auth_jwt_module.c [hunk 2]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:1]: Why ?

File: src/ngx_http_auth_jwt_module.c [hunk 3]
```diff
@@ -10,8 +11,15 @@ typedef struct {
   ngx_int_t jwt_flag;         // Function of "auth_jwt": on -> 1 | off -> 0 | $variable -> 2
   ngx_int_t jwt_var_index;    // Used only if jwt_flag==2 to fetch the $variable value
   ngx_uint_t jwt_algorithm;
+  ngx_array_t* jwt_claims;
 } ngx_http_auth_jwt_loc_conf_t;
 
+typedef struct{
+  ngx_str_t key;              //Claim Name
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:18]: Missing space

File: src/ngx_http_auth_jwt_module.c [hunk 4]
```diff
@@ -10,8 +11,15 @@ typedef struct {
...
+  ngx_str_t value;            // Claim Value
+  ngx_int_t var_index;        // if claim value was variable var_index cotains variable index. if variable index not found = -1
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:20]: Missing uppercase at the beginning of your sentence

Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:20]: Use  NGX_CONF_UNSET instead of -1

File: src/ngx_http_auth_jwt_module.c [hunk 5]
```diff
@@ -79,6 +89,12 @@ static ngx_command_t ngx_http_auth_jwt_commands[] = {
     offsetof(ngx_http_auth_jwt_loc_conf_t, jwt_algorithm),
     &ngx_http_auth_jwt_algorithms },
 
+  { ngx_string("auth_jwt_claim"),  
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE2,  
+      ngx_conf_set_auth_jwt_claim,//ngx_conf_set_keyval_slot,
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:94]: Remove commented code "ngx_conf_set_keyval_slot"

File: src/ngx_http_auth_jwt_module.c [hunk 6]
```diff
@@ -79,6 +89,12 @@ static ngx_command_t ngx_http_auth_jwt_commands[] = {
     offsetof(ngx_http_auth_jwt_loc_conf_t, jwt_algorithm),
     &ngx_http_auth_jwt_algorithms },
 
+  { ngx_string("auth_jwt_claim"),  
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE2,
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:93]: You're using 4 spaces here, the project uses 2 spaces

File: src/ngx_http_auth_jwt_module.c [hunk 7]
```diff
@@ -79,6 +89,12 @@ static ngx_command_t ngx_http_auth_jwt_commands[] = {
     offsetof(ngx_http_auth_jwt_loc_conf_t, jwt_algorithm),
     &ngx_http_auth_jwt_algorithms },
 
+  { ngx_string("auth_jwt_claim"),  
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE2,  
+      ngx_conf_set_auth_jwt_claim,//ngx_conf_set_keyval_slot,  
+      NGX_HTTP_LOC_CONF_OFFSET,  
+      offsetof(ngx_http_auth_jwt_loc_conf_t, jwt_claims),
+      NULL},
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:97]: Missing blank line after block

File: src/ngx_http_auth_jwt_module.c [hunk 8]
```diff
@@ -122,6 +138,8 @@ static ngx_int_t ngx_http_auth_jwt_handler(ngx_http_request_t *r)
 
   conf = ngx_http_get_module_loc_conf(r, ngx_http_auth_jwt_module);
 
+
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:141]: Remove this

File: src/ngx_http_auth_jwt_module.c [hunk 9]
```diff
@@ -171,10 +189,42 @@ static ngx_int_t ngx_http_auth_jwt_handler(ngx_http_request_t *r)
   time_t exp = (time_t)jwt_get_grant_int(jwt, "exp");
   if (exp != -1 && exp < time(NULL))
   {
-    ngx_log_error(NGX_LOG_INFO, r->connection->log, 0, "JWT: the jwt has expired [exp=%ld]", (long)exp);
+    ngx_log_error(NGX_LOG_WARN, r->connection->log, 0, "JWT: the jwt has expired [exp=%ld]", (long)exp);
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:192]: Revert this

File: src/ngx_http_auth_jwt_module.c [hunk 10]
```diff
@@ -171,10 +189,42 @@ static ngx_int_t ngx_http_auth_jwt_handler(ngx_http_request_t *r)
...
     return NGX_HTTP_UNAUTHORIZED;
   }
 
+  //Grant all claims
+  if(conf->jwt_claims){
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:197]: Missing space after "if"; Missign breakline after right parenthesis

File: src/ngx_http_auth_jwt_module.c [hunk 11]
```diff
@@ -208,13 +258,12 @@ static void * ngx_http_auth_jwt_create_conf(ngx_conf_t *cf)
     ngx_conf_log_error(NGX_LOG_ERR, cf, 0, "JWT: conf==NULL");
     return NULL;
   }
-
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:211]: Revert this; blank line after a block for this project

File: src/ngx_http_auth_jwt_module.c [hunk 12]
```diff
@@ -227,8 +276,8 @@ static char * ngx_http_auth_jwt_merge_conf(ngx_conf_t *cf, void *parent, void *c
   ngx_conf_merge_str_value(conf->jwt_key, prev->jwt_key, "");
   ngx_conf_merge_value(conf->jwt_var_index, prev->jwt_var_index, NGX_CONF_UNSET);
   ngx_conf_merge_value(conf->jwt_flag, prev->jwt_flag, NGX_HTTP_AUTH_JWT_OFF);
-  ngx_conf_merge_uint_value(conf->jwt_algorithm, prev->jwt_algorithm, JWT_ALG_ANY);
-
+  ngx_conf_merge_uint_value(conf->jwt_algorithm, prev->jwt_algorithm, JWT_ALG_ANY); 
+  conf->jwt_claims = !conf->jwt_claims ? prev->jwt_claims : conf->jwt_claims;
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:280]: Use ngx_conf_merge_value

File: src/ngx_http_auth_jwt_module.c [hunk 13]
```diff
@@ -450,6 +499,53 @@ static char * ngx_conf_set_auth_jwt(ngx_conf_t *cf, ngx_command_t *cmd, void *co
...
+      ngx_str_t var_name = { .data = value[2].data + 1, .len = value[2].len - 1 };
+      if(var_name.data[0] == '$'){
+        claim_conf->value = var_name;
+        claim_conf->var_index = -1;
+      }
+      else
+        claim_conf->var_index = ngx_http_get_variable_index(cf,&var_name);
+    }
+    else
+      claim_conf->var_index = -1;
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:539]: Missing block

File: src/ngx_http_auth_jwt_module.c [hunk 14]
```diff
@@ -450,6 +499,53 @@ static char * ngx_conf_set_auth_jwt(ngx_conf_t *cf, ngx_command_t *cmd, void *co
...
+    }
+
+    claim_conf = ngx_array_push(*a);
+    if (claim_conf == NULL) {
+        return NGX_CONF_ERROR;
+    }
+
+    value = cf->args->elts;
+
+    claim_conf->key= value[1];
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:526]: Missing space before =

File: src/ngx_http_auth_jwt_module.c [hunk 15]
```diff
@@ -171,10 +189,42 @@ static ngx_int_t ngx_http_auth_jwt_handler(ngx_http_request_t *r)
...
+    ngx_http_auth_jwt_claim_conf_t* claims_data = conf->jwt_claims->elts;
+    for(int i = 0; i < (int)conf->jwt_claims->nelts; i++){
+      const char* claim_name = (char*)claims_data[i].key.data;
+      const char* claim = jwt_get_grant(jwt, claim_name);
+      if(!claim){
+        ngx_log_error(NGX_LOG_WARN, r->connection->log, 0, "JWT: claim %s not found.", claims_data[i].key.data);
+        return NGX_HTTP_UNAUTHORIZED;
+      }
+
+      unsigned char* claim_value;
```
Comment (max-lt) [inline src/ngx_http_auth_jwt_module.c:207]: I think that using `ngx_str_t *` should be simpler.

**Review by max-lt — CHANGES_REQUESTED** [review]
Seems good, but there is some code style issues remaining.

File: src/ngx_http_auth_jwt_module.c [hunk 16]
```diff
@@ -523,4 +619,4 @@ static ngx_int_t auth_jwt_get_token(u_char **token, ngx_http_request_t *r, const
   }
 
   return NGX_OK;
-}
+}
```
[the remark this answers is not in the document]
Comment (aliprogrammer69) [reply src/ngx_http_auth_jwt_module.c:622]: i can't understand what is the problem

File: src/ngx_http_auth_jwt_module.c [hunk 17]
```diff
@@ -171,10 +189,42 @@ static ngx_int_t ngx_http_auth_jwt_handler(ngx_http_request_t *r)
...
+        {
+          ngx_log_error(NGX_LOG_WARN, r->connection->log, 0, "JWT: Claim Variable %s not found or empty.", claims_data[i].key.data);
+          return NGX_HTTP_UNAUTHORIZED;
+        }
+        claim_value = value->data;
+      }
+      else
+        claim_value = claims_data[i].value.data;
+
+      if(ngx_strncmp(claim,claim_value,(int)strlen(claim)) != 0){
```
[the remark this answers is not in the document]
Comment (aliprogrammer69) [reply src/ngx_http_auth_jwt_module.c:221]: because  claim is not ngx_str_t.

**Comment by max-lt** [comment]
Hi, thank you for taking the time to make the changes.
I'll review it soon, but from what I saw there may be some limitations to this additional feature (tell me if I'm wrong) :  
consider the following JWT : 

```json
{
  "sub": "1234567890",
  "name": "John Doe",
  "admin": true
}
```
so to check the suv param the config file will be 
```
auth_jwt_claim "sub" "1234567890";
```
but if I want to check to admin claim can I do 
```
auth_jwt_claim "admin" "true";
```
in other words : is the boolean parsed as string while comparing ? 

One last thing : can we compound the claims in the same block ? like
```
location /secured {
  auth_jwt_claim "sub" "1234567890";
  auth_jwt_claim "admin" "true";
}
```

**Comment by Mr-Markus** [comment]
@max-lt is this no longer maintained or why is the PR still open?
I would like to use that claim feature, but my C skills are to less to review it.

**Comment by max-lt** [comment]
I think that I will close it as it adds too much complexity to the existing code for an incomplete feature (I would expect this to work with arrays, booleans, numbers... but it only works for strings).
I would advise anyway to handle authorizations in your application, (for the moment) this module only handles authentications.

**Comment by aliprogrammer69** [comment]
@Mr-Markus the author (Max It)  asked me a question but i have forgotten to review and answer. cause i developed this feature a long time past, i have to review my codes again and then answer the "max-it". i'll review my codes and answer the question this weekend and i wish @max-lt accept and merge the PR

**Comment by Mr-Markus** [comment]
Ah ok, I understand and agree with @max-lt that complexity is growing with this feature, but I also see reasons for implementing this. 

So @max-lt, do you think you would give it a chance or not? If not, @aliprogrammer69 I would merge changes in your repo into my fork :-)

Thanks for your replies

**Comment by aliprogrammer69** [comment]
adding new features is equal to putting more complexity into the code.  
@Mr-Markus be comfortable dude, do anything you want :-)

**Comment by max-lt** [comment]
@Mr-Markus Yes of course, this is a very cool feature and it definitely have it's place here.

Here are some requirements :

## features 

### **must** have:

Simple check (on string):
```
location /secured {
  auth_jwt_claim "sub" "1234567890";
}
```

Compound check (on strings) that ensure **every** conditions is met:
```Compound check:
location /secured {
  auth_jwt_claim "sub" "1234567890";
  auth_jwt_claim "role" "admin";
}
```

### nice to have:

type casting for booleans:
```
location /secured {
  auth_jwt_claim "admin" "true::boolean";
}
```

type casting for numbers (i actually don't see any useful usage of this but why not):
```
location /secured {
  auth_jwt_claim "role" "5::number";
}
```

checks for array membership:
```
location /secured {
  # with `contains` keyword, directive would ensure that `roles` is an array
  auth_jwt_claim "roles" contains "manager";
}
```

## code seggregation 

Please move all of your additionnal code in a separate file, if this project have to grow it would be nice to be able to activate / deactivate additional feature :

```c
#include <jwt.h>
#include <jansson.h>

#define FEATURE_CLAIM
#ifdef FEATURE_CLAIM
  #include "featur_claim.h"
#endif

// ...

static ngx_command_t ngx_http_auth_jwt_commands[] = {
  // ...

#ifdef FEATURE_CLAIM
  { ngx_string("auth_jwt_claim"),  
    NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE2,  
    ngx_conf_set_auth_jwt_claim,
    NGX_HTTP_LOC_CONF_OFFSET,  
    offsetof(ngx_http_auth_jwt_loc_conf_t, jwt_claims),
    NULL },
#endif

  ngx_null_command
};

// ...
```

**Comment by max-lt** [comment]
@aliprogrammer69 actually I just reread your code and it's pretty cool as is, if you don't have time anymore for this feature I can take over from there.

**Comment by max-lt** [comment]
Ok, for now I merged you branch in a development branch ([feature/grant-claims](https://github.com/max-lt/nginx-jwt-module/tree/feature/grant-claims)), I'll thoroughly review it before merging into master.

Your branch did not passed test suite: if `auth_jwt_claim` is not set the request is crashing, fixed it in [e3877fc](https://github.com/max-lt/nginx-jwt-module/commit/c9483dd81887c90382416071a8c7f1244d4b62a7#diff-4a419c3631706839f53d27f91b97460b04fde67e4c173f6c1bd173c59d12dceaL195-R195)  (`NGX_CONF_UNSET_PTR` is not nullish)

**Comment by max-lt** [comment]
@Mr-Markus I juste released a new version (v3.2.0) that is able to perform additional verifications on JWTs claims and headers,

I finally did not used @aliprogrammer69 's solution but implemented something that looks more like [nginx solution](http://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt_require).

**Pull request merged**